## Load Python Settings<div class="tocSkip"/>



In [1]:
import sys, os
ON_COLAB = 'google.colab' in sys.modules

if ON_COLAB:
    GIT_ROOT = 'https://github.com/blueprints-for-text-analytics-python/blueprints-text/raw/master'
    os.system(f'wget {GIT_ROOT}/ch11/setup.py')

%run -i setup.py

You are working on a local system.
Files will be searched relative to "..".


In [2]:
import torch
import logging
logging.basicConfig(level=logging.ERROR)
# If there's a GPU available...
if torch.cuda.is_available():    

    # Tell PyTorch to use the GPU.    
    device = torch.device("cuda")

    print('There are %d GPU(s) available.' % torch.cuda.device_count())

    print('We will use the GPU:', torch.cuda.get_device_name(0))

# If not...
else:
    print('No GPU available, using the CPU instead.')
    device = torch.device("cpu")

There are 1 GPU(s) available.
We will use the GPU: NVIDIA GeForce RTX 3080


In [40]:
# path to import blueprints packages
#sys.path.append(BASE_DIR + '/packages')

import pandas as pd
from sklearn import preprocessing
import nltk
nltk.download('opinion_lexicon')

[nltk_data] Downloading package opinion_lexicon to
[nltk_data]     /home/setareh/nltk_data...
[nltk_data]   Package opinion_lexicon is already up-to-date!


True

# Sentiment Analysis

# Introducing the Amazon Customer Reviews Dataset

In [5]:
import pandas as pd
file = "./reviews_5_balanced.json.gz"
df = pd.read_json(file, lines=True)
df = df.drop(columns=['reviewTime','unixReviewTime']) ###
df = df.rename(columns={'reviewText': 'text'}) ###
df.sample(5, random_state=12)

overall  verified      reviewerID        asin   
163807        5     False  A2A8GHFXUG1B28  B0045Z4JAI  \
195640        5      True  A1VU337W6PKAR3  B00K0TIC56   
167820        4      True  A1Z5TT1BBSDLRM  B0012ORBT6   
104268        1     False   A4PRXX2G8900X  B005SPI45U   
51961         1      True   AYETYLNYDIS2S  B00D1HLUP8   

                                                     text   
163807  Good Decaf... it has a good flavour for a deca...  \
195640  I could not ask for a better system for my sma...   
167820  good product at a good price and saves a trip ...   
104268  I like the principle of a raw chip - something...   
51961      Fake China knockoff, you get what you pay for.   

                                                  summary  
163807                                              Nice!  
195640  I could not ask for a better system for my sma...  
167820                                         Four Stars  
104268       No better alternatives but still tastes bad.  
51961                                  Definitely not OEM

**Overall**

* This is the final rating provided by the reviewer to the product. Ranges from 1 (lowest) to 5 (highest).

**Verified**

* This indicates whether the product purchase has been verified by Amazon.

**ReviewerID**

* This is the unique identifier allocated by Amazon to each reviewer.

**ASIN**

* This is a unique product code that Amazon uses to identify the product.

**Text**

* The actual text in the review provided by the user.

**Summary**

* This is the headline or summary of the review that the user provided.


# Pre-trained Language Models using deep learning

## Deep Learning and Transfer Learning


In [6]:
# This is an optional step to reduce the size of the data by sampling only 40% of the observations

df = df.sample(frac=0.4, random_state=42)

In [7]:
df.head()

overall  verified      reviewerID        asin   
212388        5      True  A18Q3R829Q9MFS  B00UT2OFV4  \
58948         2      True  A23L2ZVK7D6LET  B00KHEJ8B6   
253525        5      True  A29P7WJ9GR4L05  B000H25VVE   
113127        1      True  A38MQ3RH7071SC  B001C8CI54   
105563        1      True   AKKQGGIXS7JBG  B00LY3TBKW   

                                                     text   
212388                                      great product  \
58948   Honestly it look really nice but I'm really ma...   
253525                                          I LOVE IT   
113127                                         wrong size   
105563                         more than half were broken   

                                                  summary  
212388                                         Five Stars  
58948   Honestly it look really nice but I'm really ma...  
253525                                         Five Stars  
113127                                           One Star  
105563                         more than half were broken

In [8]:
# Assigning a new [1,0] target class label based on the product rating
df['sentiment'] = 0
df.loc[df['overall'] > 3, 'sentiment'] = 1
df.loc[df['overall'] < 3, 'sentiment'] = 0

# Removing unecessary columns to keep a simple dataframe 
df.drop(columns=[
    'overall', 'reviewerID', 'summary', "asin"],
        inplace=True)
df.sample(3)

verified                                               text  sentiment
14077       True                     This is too thin for bed rail.          0
47921       True  it was ok, low cost, but after awhile using it...          0
184567      True                                Works as advertised          1

## Step 1: Loading models and tokenization

### Transformers library by Hugging Face is used  because of its easy-to-use functionality and wide support for multiple pretrained models.

*  Using this library, we can easily choose between different pretrained models by just changing a parameter value. The benchmark models such as BERT and GPT-2 are implement and available using this library.

* We import three classes:  
    * **config class** used to store important model parameters
    * **the tokenizer** to tokenize and prepare the text for model training
    * **model class** which defines the model architecture and weights. 
    
    
* The smallest BERT model, bert-base-uncased, which is 12 layers deep and contains 110 million parameters is used in this project.

In [9]:
from transformers import BertConfig, BertTokenizer, BertForSequenceClassification

config = BertConfig.from_pretrained('bert-base-uncased', finetuning_task='binary')
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
model = BertForSequenceClassification.from_pretrained('bert-base-uncased')

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForSequenceClassification: ['cls.predictions.transform.LayerNorm.bias', 'cls.seq_relationship.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.dense.bias', 'cls.predictions.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.dense.weight']
- This IS expected if you are initializing BertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly i

In [10]:
params = list(model.named_parameters())

print('==== Embedding Layer ====\n')

for p in params[0:5]:
    print("{:<55} {:>12}".format(p[0], str(tuple(p[1].size()))))

print('\n==== First Transformer ====\n')

for p in params[5:21]:
    print("{:<55} {:>12}".format(p[0], str(tuple(p[1].size()))))

print('\n==== Output Layer ====\n')

for p in params[-4:]:
    print("{:<55} {:>12}".format(p[0], str(tuple(p[1].size()))))


==== Embedding Layer ====

bert.embeddings.word_embeddings.weight                  (30522, 768)
bert.embeddings.position_embeddings.weight                (512, 768)
bert.embeddings.token_type_embeddings.weight                (2, 768)
bert.embeddings.LayerNorm.weight                              (768,)
bert.embeddings.LayerNorm.bias                                (768,)

==== First Transformer ====

bert.encoder.layer.0.attention.self.query.weight          (768, 768)
bert.encoder.layer.0.attention.self.query.bias                (768,)
bert.encoder.layer.0.attention.self.key.weight            (768, 768)
bert.encoder.layer.0.attention.self.key.bias                  (768,)
bert.encoder.layer.0.attention.self.value.weight          (768, 768)
bert.encoder.layer.0.attention.self.value.bias                (768,)
bert.encoder.layer.0.attention.output.dense.weight        (768, 768)
bert.encoder.layer.0.attention.output.dense.bias              (768,)
bert.encoder.layer.0.attention.output.LayerNor


*  The *tokenize* function from tokenizer class splits the sentence to tokens, pads the sentence to a fixed length sequence and output the embedding representation. 

* An attention mask is added to differentiate those positions where we have actual words from those that
contain padding characters. 



In [11]:

import warnings; ###
warnings.filterwarnings('ignore'); ###

def get_tokens(text, tokenizer, max_seq_length, add_special_tokens=True):
  input_ids = tokenizer.encode(text, 
                               add_special_tokens=add_special_tokens, 
                               max_length=max_seq_length,
                               pad_to_max_length=True,
                              truncation=True)
  attention_mask = [int(id > 0) for id in input_ids]
  assert len(input_ids) == max_seq_length
  assert len(attention_mask) == max_seq_length
  return (input_ids, attention_mask)

text = "Here is the sentence I want embeddings for."
input_ids, attention_mask = get_tokens(text, 
                                       tokenizer, 
                                       max_seq_length=30, 
                                       add_special_tokens = True)
input_tokens = tokenizer.convert_ids_to_tokens(input_ids)
print (text)
print (input_tokens)
print (input_ids)
print (attention_mask)

Here is the sentence I want embeddings for.
['[CLS]', 'here', 'is', 'the', 'sentence', 'i', 'want', 'em', '##bed', '##ding', '##s', 'for', '.', '[SEP]', '[PAD]', '[PAD]', '[PAD]', '[PAD]', '[PAD]', '[PAD]', '[PAD]', '[PAD]', '[PAD]', '[PAD]', '[PAD]', '[PAD]', '[PAD]', '[PAD]', '[PAD]', '[PAD]']
[101, 2182, 2003, 1996, 6251, 1045, 2215, 7861, 8270, 4667, 2015, 2005, 1012, 102, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]


### [CLS] token, which stands for classification, which is one of the pretraining tasks of the BERT model. This token is used to identify the start of a sentence and stores the aggregated representation of the entire sentence.

###  The ## is used to identify tokens that are subwords, which is a special characteristic of the BERT model. It is used to distinct between root words, prefixes, and suffixes

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(df['text'],
                                                    df['sentiment'],
                                                    test_size=0.2,
                                                    random_state=42,
                                                    stratify=df['sentiment'])
X_train_tokens = X_train.apply(get_tokens, args=(tokenizer, 50))
X_test_tokens = X_test.apply(get_tokens, args=(tokenizer, 50))

In [13]:
import torch
from torch.utils.data import TensorDataset

input_ids_train = torch.tensor(
    [features[0] for features in X_train_tokens.values], dtype=torch.long)
input_mask_train = torch.tensor(
    [features[1] for features in X_train_tokens.values], dtype=torch.long)
label_ids_train = torch.tensor(Y_train.values, dtype=torch.long)

print (input_ids_train.shape)
print (input_mask_train.shape)
print (label_ids_train.shape)

torch.Size([94156, 50])
torch.Size([94156, 50])
torch.Size([94156])


The token tensor contains a mapping to the BERT vocabulary. The number 101 indicates the start, and 102 indicates the end of the review sentence. 

In [14]:
input_ids_train[2]

tensor([ 101, 6581, 3737,  102,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0])

We combine these tensors into a TensorDataset, which is the basic data structure used to load all observations during model training:

In [15]:
train_dataset = TensorDataset(input_ids_train,input_mask_train,label_ids_train)

In [16]:
input_ids_test = torch.tensor([features[0] for features in X_test_tokens.values], 
                              dtype=torch.long)
input_mask_test = torch.tensor([features[1] for features in X_test_tokens.values], 
                               dtype=torch.long)
label_ids_test = torch.tensor(Y_test.values, 
                              dtype=torch.long)
test_dataset = TensorDataset(input_ids_test, input_mask_test, label_ids_test)

## Step 2 - Model Training

In [17]:
from torch.utils.data import DataLoader, RandomSampler

train_batch_size = 64
num_train_epochs = 2

train_sampler = RandomSampler(train_dataset)
train_dataloader = DataLoader(train_dataset, 
                              sampler=train_sampler, 
                              batch_size=train_batch_size)
t_total = len(train_dataloader) // num_train_epochs

print ("Num examples = ", len(train_dataset))
print ("Num Epochs = ", num_train_epochs)
print ("Total train batch size  = ", train_batch_size)
print ("Total optimization steps = ", t_total)

Num examples =  94156
Num Epochs =  2
Total train batch size  =  64
Total optimization steps =  736


These parameters are used based on the parameters in the BERT paper and recommendations in the Transformers library,

In [22]:
import torch_optimizer as optim

from transformers import AdamW , get_linear_schedule_with_warmup

learning_rate = 1e-4
adam_epsilon = 1e-8
warmup_steps = 0

optimizer = AdamW(model.parameters(), lr=learning_rate, eps=adam_epsilon)
scheduler = get_linear_schedule_with_warmup(optimizer, 
                                            num_warmup_steps=warmup_steps, 
                                            num_training_steps=t_total)

In [23]:
from tqdm import trange, notebook

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
train_iterator = trange(num_train_epochs, desc="Epoch")

## Put model in 'train' mode
model.train()
    
for epoch in train_iterator:
    epoch_iterator = notebook.tqdm(train_dataloader, desc="Iteration")
    for step, batch in enumerate(epoch_iterator):

        ## Reset all gradients at start of every iteration
        model.zero_grad()
        
        ## Put the model and the input observations to GPU
        model.to(device)
        batch = tuple(t.to(device) for t in batch)
        
        ## Identify the inputs to the model
        inputs = {'input_ids':      batch[0],
                  'attention_mask': batch[1],
                  'labels':         batch[2]}

        ## Forward Pass through the model. Input -> Model -> Output
        outputs = model(**inputs)

        ## Determine the deviation (loss)
        loss = outputs[0]
        print("\r%f" % loss, end='')

        ## Back-propogate the loss (automatically calculates gradients)
        loss.backward()

        ## Prevent exploding gradients by limiting gradients to 1.0 
        torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)

        ## Update the parameters and learning rate
        optimizer.step()
        scheduler.step()

Epoch:   0%|          | 0/2 [00:00<?, ?it/s]

Iteration:   0%|          | 0/1472 [00:00<?, ?it/s]

0.247789

Epoch:  50%|█████     | 1/2 [03:50<03:50, 230.11s/it]

0.015899

Iteration:   0%|          | 0/1472 [00:00<?, ?it/s]

0.265361

Epoch: 100%|██████████| 2/2 [07:38<00:00, 229.18s/it]

0.011472

In [24]:
model.save_pretrained('outputs')

## Step 3 - Model Evaluation


In [26]:
import numpy as np
from sklearn.metrics import accuracy_score

from torch.utils.data import SequentialSampler

test_batch_size = 64
test_sampler = SequentialSampler(test_dataset)
test_dataloader = DataLoader(test_dataset, 
                             sampler=test_sampler, 
                             batch_size=test_batch_size)

# Load the pre-trained model that was saved earlier 
# model = model.from_pretrained('/outputs')

# Initialize the prediction and actual labels
preds = None
out_label_ids = None

## Put model in "eval" mode
model.eval()

for batch in notebook.tqdm(test_dataloader, desc="Evaluating"):
    
    ## Put the model and the input observations to GPU
    model.to(device)
    batch = tuple(t.to(device) for t in batch)
    
    ## Do not track any gradients since in 'eval' mode
    with torch.no_grad():
        inputs = {'input_ids':      batch[0],
                  'attention_mask': batch[1],
                  'labels':         batch[2]}

        ## Forward pass through the model
        outputs = model(**inputs)

        ## We get loss since we provided the labels
        tmp_eval_loss, logits = outputs[:2]

        ## There maybe more than one batch of items in the test dataset
        if preds is None:
            preds = logits.detach().cpu().numpy()
            out_label_ids = inputs['labels'].detach().cpu().numpy()
        else:
            preds = np.append(preds, logits.detach().cpu().numpy(), axis=0)
            out_label_ids = np.append(out_label_ids, 
                                      inputs['labels'].detach().cpu().numpy(), 
                                      axis=0)
    
## Get final loss, predictions and accuracy
preds = np.argmax(preds, axis=1)
acc_score = accuracy_score(preds, out_label_ids)
print ('Accuracy Score on Test data ', acc_score)

Evaluating:   0%|          | 0/368 [00:00<?, ?it/s]

Accuracy Score on Test data  0.949660152931181
